### DISCUSSION TITLE
Webcil reading via SRM or other public library

### DISCUSSION THREAD

**Discussion by christophwille** [dotnet/runtime#86587] [General]
https://github.com/dotnet/runtime/blob/main/docs/design/mono/webcil.md

I found https://github.com/dotnet/runtime/tree/main/src/libraries/Microsoft.NET.WebAssembly.Webcil/src/Webcil - will that be turned into an official Webcil-reading library for NuGet or merged with System.Reflection.Metadata (SRM) or are there no plans for an official, public library to read Webcil?

Asking for a friend ;-) I work on ILSpy and we use SRM to read bog-standard assemblies.

**Comment by jkotas** [comment]
cc @lambdageek

**Comment by lambdageek** [comment]
> will that be turned into an official Webcil-reading library for NuGet or merged with System.Reflection.Metadata (SRM) or are there no plans for an official, public library to read Webcil?

That's a good question.

On the one hand, in some ways Webcil is an internal implementation detail of .NET for WebAssembly.  The same way that, for example, NativeAOT's metadata representation or a single-file .NET Core binary's implementation of assemblies  is not really a part of its public interface - you can probably poke around in a binary image and find something that represents a Type. Webcil is an artifact of the deployment process - it just happens that at the moment there's a clear 1-1 relationship between individual assemblies and the .wasm modules.  The main reason I made a versioned spec for the file format is because we want anti-virus tools to understand the files.

On the other hand, I'm sympathetic to the users will want to run other tools to examine the files.  I don't think it's reasonable to try and stop them. 

I'm not sure if we would make the library public in the .NET 8 timeframe - depends on whether we get good feedback in the .NET 8 previews that Webcil is really solving problems for customers.

I would love to make WebcilReader part of SRM, but there's an annoying technical complication: SRM is used by Visual Studio and the version of MSBuild that ships with it (that is targeting .NET Framework).  So if we make WebcilReader part of SRM, we won't actually get to use it in our build tasks for the .NET for WebAssembly SDK until VS bumps their SRM dependency.

If we make a separate Webcil reader nuget I don't see us having the same problems, but we also won't get to use the nice SRM internal classes in the implementation of webcil reader, which is a shame (right now the library does a lot more array allocation and copying than I would have liked)

TL;DR I don't want to keep the webcil reader dll private to the .NET SDK, but I also don't have a schedule for how to make it public.

**Reply by christophwille** [reply]
We are even today using a non-standard library in R2R, namely ILCompiler.Reflection.ReadyToRun.Experimental (see https://github.com/icsharpcode/ILSpy/blob/220b4cbd0675c3d1c5a7b26e9fa3ba4efb728ccd/ILSpy.ReadyToRun/ILSpy.ReadyToRun.csproj#L41) from dotnet8-transport (see https://github.com/icsharpcode/ILSpy/blob/220b4cbd0675c3d1c5a7b26e9fa3ba4efb728ccd/NuGet.config#L6). 

Having such a non-public non-supported way of getting access to Microsoft.NET.WebAssembly.Webcil could be a crutch to get started - because we cannot push it down to the decompilation engine project as it is consumed downstream by projects that most likely cannot use a custom feed (we'd have to eg have a callback "Cannot read file, can your callback?" inside ILSpy, nasty...). 

Copying code would be the worst approach I could think of.

**Reply by lambdageek** [reply]
> Having such a non-public non-supported way of getting access to Microsoft.NET.WebAssembly.Webcil could be a crutch to get started

I think we can probably do a transport nuget

**Reply by lambdageek** [reply]
https://github.com/dotnet/runtime/issues/86597

**Reply by StefH** [reply]
@lambdageek 

Is it possible to use the `WebcilReader`-class from that Microsoft.NET.WebAssembly.Webcil project to read a **Webcil - wasm** file en extract the PE bytes?

I want to create a `MetadataReference` based on the PE bytes, this is needed to compile a CSharpSyntaxTree using CSharpCompilation.Create(...).

**Reply by lambdageek** [reply]
>  extract the PE bytes?

@StefH There are no PE bytes in a webcil file. a Webcil file contains COFF sections that contain ECMA 335 metadata tables and heaps, but not other PE data.  It maybe possible to do some kind of reverse synthesis to create something that looks like a PE image that contains the same ECMA 335 tables and heaps, but it won't be byte-for-byte identical to the starting PE image.  But if you just need to give Roslyn something it can work with, that might be good enough.

The best approach would be to create a new subclass of `MetadataReference` that returns a new subclass of `Metadata` (like [`AssemblyMetadata`](https://github.com/dotnet/roslyn/blob/bfbb6dca6464fb27c325338ce408ce08f516309b/src/Compilers/Core/Portable/MetadataReference/AssemblyMetadata.cs) that uses the `WebcilReader` instead of a `PEReader` to get at the underlying CLI image). But I'm pretty sure that won't work - it doesn't look like this part of Roslyn is meant to be extensible by external code.  (In particular there doesn't seem to be a good way to create a [`PEModule`](https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/MetadataReader/PEModule.cs) that is reading from something other than a `PEReader`).

**Reply by StefH** [reply]
In case anyone is interested:

I create a project which can create a valid `MetadataReference` based on the **Webcil WASM wrapped** file : https://github.com/StefH/ProtoBufJsonConverter/tree/main/src-webcil
